File: backend/routers/teachers.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import distinct, func, select
from sqlalchemy.orm import Session

from .. import models, schemas
from ..db import get_db
from ..deps import current_teacher

router = APIRouter(prefix="/api/teachers", tags=["teachers"])


def _ensure_self_or_admin(teacher: models.Teacher, target_id: int) -> None:
    if teacher.teacher_id != target_id and teacher.role != "admin":
        raise HTTPException(403, "you can only access your own data")


@router.get("/{teacher_id}/dashboard", response_model=schemas.DashboardOut)
def dashboard(teacher_id: int, db: Session = Depends(get_db), me: models.Teacher = Depends(current_teacher)) -> schemas.DashboardOut:
    _ensure_self_or_admin(me, teacher_id)
    teacher = db.get(models.Teacher, teacher_id)
    if not teacher:
        raise HTTPException(404, "teacher not found")
    institution = teacher.institution

    student_count = db.execute(
        select(func.count(distinct(models.Enrollment.student_id)))
        .join(models.Class, models.Class.class_id == models.Enrollment.class_id)
        .where(models.Class.teacher_id == teacher_id)
    ).scalar_one()

    cluster_count = db.execute(
        select(func.count(distinct(models.Student.cluster_id)))
        .join(models.Enrollment, models.Enrollment.student_id == models.Student.student_id)
        .join(models.Class, models.Class.class_id == models.Enrollment.class_id)
        .where(models.Class.teacher_id == teacher_id)
    ).scalar_one()

    adaptation_count = db.execute(
        select(func.count(models.AdaptedLesson.adapted_id)).where(models.AdaptedLesson.teacher_id == teacher_id)
    ).scalar_one()

    kb_count = db.execute(select(func.count(models.KnowledgeBase.kb_id))).scalar_one()

    # Recent adapted lessons with their head version number
    head_version_subq = (
        select(
            models.LessonPlanVersion.adapted_id,
            func.max(models.LessonPlanVersion.version_number).label("head_n"),
        )
        .where(models.LessonPlanVersion.is_head == 1)
        .group_by(models.LessonPlanVersion.adapted_id)
        .subquery()
    )

    recent_rows = db.execute(
        select(
            models.AdaptedLesson,
            models.Lesson.title,
            models.Lesson.grade_level,
            models.Lesson.cs_topic,
            models.StudentCluster.cluster_name,
            func.coalesce(head_version_subq.c.head_n, 1).label("head_n"),
        )
        .join(models.Lesson, models.Lesson.lesson_id == models.AdaptedLesson.lesson_id)
        .join(models.StudentCluster, models.StudentCluster.cluster_id == models.AdaptedLesson.cluster_id)
        .outerjoin(head_version_subq, head_version_subq.c.adapted_id == models.AdaptedLesson.adapted_id)
        .where(models.AdaptedLesson.teacher_id == teacher_id)
        .order_by(models.AdaptedLesson.generated_at.desc())
        .limit(6)
    ).all()

    recent = [
        schemas.RecentAdaptation(
            adapted_id=al.adapted_id,
            lesson_title=title,
            grade_level=grade,
            cs_topic=topic,
            cluster_name=cn,
            head_version_number=int(hn),
            generated_at=al.generated_at,
        )
        for al, title, grade, topic, cn, hn in recent_rows
    ]

    roster_rows = db.execute(
        select(
            models.Student.student_id,
            models.Student.first_name,
            models.Student.last_name,
            models.StudentCluster.cluster_name,
            models.Class.class_name,
        )
        .join(models.Enrollment, models.Enrollment.student_id == models.Student.student_id)
        .join(models.Class, models.Class.class_id == models.Enrollment.class_id)
        .outerjoin(models.StudentCluster, models.StudentCluster.cluster_id == models.Student.cluster_id)
        .where(models.Class.teacher_id == teacher_id)
        .order_by(models.Class.class_name, models.Student.last_name)
    ).all()

    roster = [
        schemas.RosterEntry(
            student_id=sid,
            student_name=f"{fn} {ln}",
            cluster_name=cn,
            class_name=cln,
        )
        for sid, fn, ln, cn, cln in roster_rows
    ]

    return schemas.DashboardOut(
        teacher=schemas.TeacherOut.model_validate(teacher),
        institution=schemas.InstitutionOut.model_validate(institution) if institution else None,
        metrics={
            "students": int(student_count),
            "clusters": int(cluster_count),
            "adaptations": int(adaptation_count),
            "knowledge_bases": int(kb_count),
            "classes": len(set(r.class_name for r in roster)),
        },
        recent_adaptations=recent,
        roster=roster,
    )


@router.get("/{teacher_id}/classes", response_model=list[schemas.ClassOut])
def classes(teacher_id: int, db: Session = Depends(get_db), me: models.Teacher = Depends(current_teacher)) -> list[schemas.ClassOut]:
    _ensure_self_or_admin(me, teacher_id)
    classes = list(db.execute(select(models.Class).where(models.Class.teacher_id == teacher_id)).scalars())
    out: list[schemas.ClassOut] = []
    for cl in classes:
        rows = db.execute(
            select(models.Student, models.StudentCluster.cluster_name)
            .join(models.Enrollment, models.Enrollment.student_id == models.Student.student_id)
            .outerjoin(models.StudentCluster, models.StudentCluster.cluster_id == models.Student.cluster_id)
            .where(models.Enrollment.class_id == cl.class_id)
            .order_by(models.Student.last_name)
        ).all()
        students = [
            schemas.StudentOut(
                student_id=s.student_id,
                first_name=s.first_name,
                last_name=s.last_name,
                cluster_id=s.cluster_id,
                cluster_name=cn,
                math_performance=s.math_performance,
                ela_performance=s.ela_performance,
                learner_variability=s.learner_variability,
            )
            for s, cn in rows
        ]
        out.append(
            schemas.ClassOut(
                class_id=cl.class_id,
                class_name=cl.class_name,
                grade_band=cl.grade_band,
                subject=cl.subject,
                school_year=cl.school_year,
                students=students,
            )
        )
    return out


@router.patch("/{teacher_id}/students/{student_id}", response_model=schemas.StudentOut)
def update_student(
    teacher_id: int,
    student_id: int,
    body: schemas.StudentUpdateIn,
    db: Session = Depends(get_db),
    me: models.Teacher = Depends(current_teacher),
) -> schemas.StudentOut:
    _ensure_self_or_admin(me, teacher_id)
    student = db.execute(
        select(models.Student)
        .join(models.Enrollment, models.Enrollment.student_id == models.Student.student_id)
        .join(models.Class, models.Class.class_id == models.Enrollment.class_id)
        .where(models.Class.teacher_id == teacher_id, models.Student.student_id == student_id)
        .limit(1)
    ).scalar_one_or_none()
    if not student:
        raise HTTPException(404, "student not found for this teacher")

    if body.cluster_id is not None:
        cluster = db.get(models.StudentCluster, body.cluster_id)
        if not cluster:
            raise HTTPException(404, "cluster not found")
        student.cluster_id = body.cluster_id
    if body.math_performance is not None:
        student.math_performance = body.math_performance
    if body.ela_performance is not None:
        student.ela_performance = body.ela_performance
    if body.learner_variability is not None:
        student.learner_variability = body.learner_variability

    db.commit()
    db.refresh(student)
    cluster_name = student.cluster.cluster_name if student.cluster else None
    return schemas.StudentOut(
        student_id=student.student_id,
        first_name=student.first_name,
        last_name=student.last_name,
        cluster_id=student.cluster_id,
        cluster_name=cluster_name,
        math_performance=student.math_performance,
        ela_performance=student.ela_performance,
        learner_variability=student.learner_variability,
    )

File: server/src/routes/teachers.js
const express = require('express');
const db = require('../db');
const requireAuth = require('../middleware/auth');
const { requireOwnerOrAdmin } = require('../middleware/rbac');
const { NotFoundError, ValidationError } = require('../errors');

const router = express.Router();

// GET /api/teachers/:id/dashboard — full Python parity (TEACH-01)
router.get('/:id/dashboard', requireAuth, requireOwnerOrAdmin, (req, res) => {
  const teacherId = parseInt(req.params.id);

  // Fetch teacher
  const teacher = db.prepare(
    'SELECT teacher_id, first_name, last_name, email, institution_id, role FROM teacher WHERE teacher_id = ?'
  ).get(teacherId);

  if (!teacher) {
    throw new NotFoundError('Teacher');
  }

  // Fetch institution (if exists)
  let institution = null;
  if (teacher.institution_id) {
    institution = db.prepare(
      'SELECT institution_id, name, type, district FROM institution WHERE institution_id = ?'
    ).get(teacher.institution_id);
  }

  // Count students
  const { student_count } = db.prepare(
    'SELECT COUNT(DISTINCT e.student_id) as student_count FROM enrollment e JOIN class c ON e.class_id = c.class_id WHERE c.teacher_id = ?'
  ).get(teacherId);

  // Count clusters
  const { cluster_count } = db.prepare(
    'SELECT COUNT(DISTINCT s.cluster_id) as cluster_count FROM student s JOIN enrollment e ON s.student_id = e.student_id JOIN class c ON e.class_id = c.class_id WHERE c.teacher_id = ?'
  ).get(teacherId);

  // Count adaptations
  const { adaptation_count } = db.prepare(
    'SELECT COUNT(*) as adaptation_count FROM adapted_lesson WHERE teacher_id = ?'
  ).get(teacherId);

  // Count KBs (global)
  const { kb_count } = db.prepare('SELECT COUNT(*) as kb_count FROM knowledge_base').get();

  // Recent adaptations (last 6)
  const recentRows = db.prepare(`
    SELECT al.adapted_id, l.title as lesson_title, l.grade_level, l.cs_topic,
           sc.cluster_name, al.generated_at
    FROM adapted_lesson al
    JOIN lesson l ON al.lesson_id = l.lesson_id
    JOIN student_cluster sc ON sc.cluster_id = al.cluster_id
    WHERE al.teacher_id = ?
    ORDER BY al.generated_at DESC
    LIMIT 6
  `).all(teacherId);

  // Roster (all students in teacher's classes)
  const rosterRows = db.prepare(`
    SELECT s.student_id, s.first_name, s.last_name,
           sc.cluster_name, c.class_name
    FROM student s
    JOIN enrollment e ON s.student_id = e.student_id
    JOIN class c ON e.class_id = c.class_id
    LEFT JOIN student_cluster sc ON s.cluster_id = sc.cluster_id
    WHERE c.teacher_id = ?
    ORDER BY c.class_name, s.last_name
  `).all(teacherId);

  const roster = rosterRows.map(r => ({
    student_id: r.student_id,
    student_name: `${r.first_name} ${r.last_name}`,
    cluster_name: r.cluster_name,
    class_name: r.class_name
  }));

  const classesCount = [...new Set(rosterRows.map(r => r.class_name))].length;

  res.json({
    teacher,
    institution,
    metrics: {
      students: student_count,
      clusters: cluster_count,
      adaptations: adaptation_count,
      knowledge_bases: kb_count,
      classes: classesCount
    },
    recent_adaptations: recentRows,
    roster
  });
});

// GET /api/teachers/:id/classes — classes with nested students (TEACH-02, TEACH-05)
router.get('/:id/classes', requireAuth, requireOwnerOrAdmin, (req, res) => {
  const teacherId = parseInt(req.params.id);

  // Fetch classes
  const classes = db.prepare(
    'SELECT class_id, class_name, grade_band, subject, school_year FROM class WHERE teacher_id = ? ORDER BY class_id'
  ).all(teacherId);

  if (classes.length === 0) {
    return res.json([]);
  }

  // Fetch all students for all classes in a single JOIN query (avoid N+1)
  const classIds = classes.map(c => c.class_id);
  const placeholders = classIds.map(() => '?').join(',');
  const studentRows = db.prepare(`
    SELECT s.student_id, s.first_name, s.last_name, s.cluster_id,
           sc.cluster_name, s.math_performance, s.ela_performance, s.learner_variability,
           e.class_id
    FROM student s
    JOIN enrollment e ON s.student_id = e.student_id
    LEFT JOIN student_cluster sc ON s.cluster_id = sc.cluster_id
    WHERE e.class_id IN (${placeholders})
    ORDER BY e.class_id, s.last_name
  `).all(...classIds);

  // Group students by class_id
  const studentsByClass = {};
  for (const row of studentRows) {
    if (!studentsByClass[row.class_id]) {
      studentsByClass[row.class_id] = [];
    }
    studentsByClass[row.class_id].push({
      student_id: row.student_id,
      first_name: row.first_name,
      last_name: row.last_name,
      cluster_id: row.cluster_id,
      cluster_name: row.cluster_name,
      math_performance: row.math_performance,
      ela_performance: row.ela_performance,
      learner_variability: row.learner_variability
    });
  }

  // Attach students to each class
  const classesWithStudents = classes.map(c => ({
    class_id: c.class_id,
    class_name: c.class_name,
    grade_band: c.grade_band,
    subject: c.subject,
    school_year: c.school_year,
    students: studentsByClass[c.class_id] || []
  }));

  res.json(classesWithStudents);
});

// PATCH /api/teachers/:id/students/:student_id — update student cluster (TEACH-03, CLUS-04)
router.patch('/:id/students/:student_id', requireAuth, requireOwnerOrAdmin, (req, res) => {
  const teacherId = parseInt(req.params.id);
  const studentId = parseInt(req.params.student_id);

  // Verify student belongs to teacher
  const student = db.prepare(`
    SELECT s.student_id, s.first_name, s.last_name, s.cluster_id,
           s.math_performance, s.ela_performance, s.learner_variability
    FROM student s
    JOIN enrollment e ON s.student_id = e.student_id
    JOIN class c ON e.class_id = c.class_id
    WHERE c.teacher_id = ? AND s.student_id = ?
  `).get(teacherId, studentId);

  if (!student) {
    throw new NotFoundError('Student');
  }

  const { cluster_id, math_performance, ela_performance, learner_variability } = req.body;

  // If cluster_id provided, verify cluster exists
  if (cluster_id !== undefined && cluster_id !== null) {
    const cluster = db.prepare(
      'SELECT cluster_id FROM student_cluster WHERE cluster_id = ?'
    ).get(cluster_id);

    if (!cluster) {
      throw new NotFoundError('Cluster');
    }
  }

  // Build UPDATE dynamically (only non-null fields)
  const updates = [];
  const values = [];

  if (cluster_id !== undefined && cluster_id !== null) {
    updates.push('cluster_id = ?');
    values.push(cluster_id);
  }
  if (math_performance !== undefined && math_performance !== null) {
    updates.push('math_performance = ?');
    values.push(math_performance);
  }
  if (ela_performance !== undefined && ela_performance !== null) {
    updates.push('ela_performance = ?');
    values.push(ela_performance);
  }
  if (learner_variability !== undefined && learner_variability !== null) {
    updates.push('learner_variability = ?');
    values.push(learner_variability);
  }

  if (updates.length > 0) {
    values.push(studentId);
    db.prepare(`UPDATE student SET ${updates.join(', ')} WHERE student_id = ?`).run(...values);
  }

  // Fetch updated student with cluster_name
  const updated = db.prepare(`
    SELECT s.student_id, s.first_name, s.last_name, s.cluster_id,
           sc.cluster_name, s.math_performance, s.ela_performance, s.learner_variability
    FROM student s
    LEFT JOIN student_cluster sc ON s.cluster_id = sc.cluster_id
    WHERE s.student_id = ?
  `).get(studentId);

  res.json(updated);
});

// GET /api/teachers/:id/profile — view profile (TEACH-04)
router.get('/:id/profile', requireAuth, requireOwnerOrAdmin, (req, res) => {
  const teacherId = parseInt(req.params.id);

  const teacher = db.prepare(
    'SELECT teacher_id, first_name, last_name, email, role, institution_id FROM teacher WHERE teacher_id = ?'
  ).get(teacherId);

  if (!teacher) {
    throw new NotFoundError('Teacher');
  }

  res.json(teacher);
});

// PUT /api/teachers/:id/profile — update profile name only (TEACH-04)
router.put('/:id/profile', requireAuth, requireOwnerOrAdmin, (req, res) => {
  const teacherId = parseInt(req.params.id);

  // Verify teacher exists
  const teacher = db.prepare(
    'SELECT teacher_id, first_name, last_name FROM teacher WHERE teacher_id = ?'
  ).get(teacherId);

  if (!teacher) {
    throw new NotFoundError('Teacher');
  }

  const { first_name, last_name } = req.body;

  // Validate: non-empty strings, max 50 chars
  if (!first_name || typeof first_name !== 'string' || first_name.length === 0 || first_name.length > 50) {
    throw new ValidationError('first_name must be a non-empty string (max 50 chars)');
  }
  if (!last_name || typeof last_name !== 'string' || last_name.length === 0 || last_name.length > 50) {
    throw new ValidationError('last_name must be a non-empty string (max 50 chars)');
  }

  // Update name fields only (NO email or role changes)
  db.prepare(
    'UPDATE teacher SET first_name = ?, last_name = ? WHERE teacher_id = ?'
  ).run(first_name, last_name, teacherId);

  // Fetch updated teacher
  const updated = db.prepare(
    'SELECT teacher_id, first_name, last_name, email, role, institution_id FROM teacher WHERE teacher_id = ?'
  ).get(teacherId);

  res.json(updated);
});

module.exports = router;
